powa--3.1.2--3.2.0dev.sql

```diff
@@ -1,6 +1,388 @@
+    OUT relid oid,
+    OUT numscan bigint,
+    OUT tup_returned bigint,
+    OUT tup_fetched bigint,
+    OUT n_tup_ins bigint,
+    OUT n_tup_upd bigint,
+    OUT n_tup_del bigint,
+    OUT n_tup_hot_upd bigint,
+    OUT n_liv_tup bigint,
+    OUT n_dead_tup bigint,
+    OUT n_mod_since_analyze bigint,
+    OUT blks_read bigint,
+    OUT blks_hit bigint,
+    OUT last_vacuum timestamp with time zone,
+    OUT vacuum_count bigint,
+    OUT last_autovacuum timestamp with time zone,
+    OUT autovacuum_count bigint,
+    OUT last_analyze timestamp with time zone,
+    OUT analyze_count bigint,
+    OUT last_autoanalyze timestamp with time zone,
+    OUT autoanalyze_count bigint)
+RETURNS SETOF record AS $PROC$
+       SELECT d.oid AS dbid, r.relid, r.numscan, r.tup_returned, r.tup_fetched,
+       r.n_tup_ins, r.n_tup_upd, r.n_tup_del, r.n_tup_hot_upd, r.n_liv_tup,
+       r.n_dead_tup, r.n_mod_since_analyze, r.blks_read, r.blks_hit, r.last_vacuum,
+       r.vacuum_count, r.last_autovacuum, r.autovacuum_count, r.last_analyze,
+       r.analyze_count, r.last_autoanalyze, r.autoanalyze_count
+       FROM pg_database d, powa_stat_all_rel(d.oid) as r
+$PROC$ LANGUAGE sql; /* end of get_all_relations_capture_data */
+
+CREATE OR REPLACE FUNCTION powa_all_relations_snapshot() RETURNS void AS $PROC$
+DECLARE
+    result boolean;
+    v_funcname    text := 'powa_all_relations_snapshot';
+    v_rowcount    bigint;
+BEGIN
+    PERFORM powa_log(format('running %I', v_funcname));
+
+    -- Insert cluster-wide relation statistics
+    WITH rel AS (
+        SELECT * FROM get_all_relations_capture_data()
+    )
+    INSERT INTO powa_all_relations_history_current
+        SELECT dbid, relid,
+        ROW(now(), numscan, tup_returned, tup_fetched,
+            n_tup_ins, n_tup_upd, n_tup_del, n_tup_hot_upd,
+            n_liv_tup, n_dead_tup, n_mod_since_analyze,
+            blks_read, blks_hit, last_vacuum, vacuum_count,
+            last_autovacuum, autovacuum_count, last_analyze,
+            analyze_count, last_autoanalyze,
+            autoanalyze_count)::powa_all_relations_history_record AS record
+        FROM rel;
+
+    GET DIAGNOSTICS v_rowcount = ROW_COUNT;
+
+    perform powa_log(format('%I - rowcount: %s',
+            v_funcname, v_rowcount));
+
+    result := true;
+END;
+$PROC$ language plpgsql; /* end of powa_all_relations_snapshot */
+
+
+CREATE OR REPLACE FUNCTION get_kcache_capture_data(OUT queryid bigint, OUT userid oid, OUT dbid oid,
+OUT reads bigint, OUT writes bigint, OUT user_time double precision, OUT system_time double precision)
+RETURNS SETOF record AS $PROC$
+BEGIN
+        RETURN QUERY
+        SELECT k.queryid, k.userid, k.dbid, k.reads, k.writes, k.user_time, k.system_time
+        FROM pg_stat_kcache() k
+        JOIN pg_roles r ON r.oid = k.userid
+        WHERE NOT (r.rolname = ANY (string_to_array(current_setting('powa.ignored_users'),',')));
+END;
+$PROC$ LANGUAGE plpgsql; /* end of get_kcache_capture_data() */
```

Comment:
Why using plpgsql instead of sql here and some other functions?  If it's to avoid an error at creation time, you could wrap the code in

```
SET check_function_bodies = off;
[...]
RESET check_function_bodies;
```

I think that using a plpgsql function will internally materialize the result set, which itself should also be materialize due to the CTE, so it's better to use plain SQL fuctions?

Reply:
Yeah a SQL function would probably be better here